controller/volume_rebuilding_controller.go

```diff
@@ -191,13 +318,122 @@ func (vbc *VolumeRebuildingController) reconcile(volName string) (err error) {
+	}
+
+	if !isOfflineRebuildEnabled && util.IsHigherPriorityVATicketExisting(va, longhorn.AttacherTypeVolumeRebuildingController) {
+		warnMsg := fmt.Sprintf("Canceling volume %v offline rebuilding", volName)
+		vbc.logger.Info(warnMsg)
+		vbc.eventRecorder.Event(vol, corev1.EventTypeWarning, constant.EventReasonCanceledOfflineRebuilding, warnMsg)
+	}
+
+	delete(va.Spec.AttachmentTickets, rebuildingAttachmentTicketID)
 	return nil
 }
 
+func (vbc *VolumeRebuildingController) isVolumeOfflineRebuildEnabled(offlineRebuilding longhorn.VolumeOfflineRebuilding) (bool, error) {
+	offlineReplicaRebuildingValue, err := vbc.ds.GetSettingAsBool(types.SettingNameOfflineReplicaRebuilding)
+	if err != nil {
+		return false, err
+	}
+	return (offlineRebuilding == longhorn.VolumeOfflineRebuildingEnabled) || (offlineRebuilding == longhorn.VolumeOfflineRebuildingIgnored && offlineReplicaRebuildingValue), nil
+}
+
+func (vbc *VolumeRebuildingController) syncLHVolumeAttachementForOfflineRebuild(vol *longhorn.Volume, va *longhorn.VolumeAttachment, attachmentID string) (*longhorn.VolumeAttachment, error) {
+	replicas, err := vbc.ds.ListVolumeReplicasRO(vol.Name)
+	if err != nil {
+		return va, err
+	}
+	areAllReplicasHealthy, err := vbc.isVolumeReplicasHealthy(vol.Spec.NumberOfReplicas, replicas)
+	if err != nil {
+		return va, err
+	}
+	if !areAllReplicasHealthy {
+		createOrUpdateAttachmentTicket(va, attachmentID, vol.Status.OwnerID, longhorn.TrueValue, longhorn.AttacherTypeVolumeRebuildingController)
+	}
+	return va, nil
+}
+
+func (vbc *VolumeRebuildingController) isVolumeReplicasHealthy(numberOfReplicas int, replicas map[string]*longhorn.Replica) (bool, error) {
+	healthyCount := 0
+	for _, replica := range replicas {
+		node, err := vbc.ds.GetNodeRO(replica.Spec.NodeID)
+		if err != nil {
+			if apierrors.IsNotFound(err) {
+				continue
+			}
+			return false, err
+		}
+		isNodeReady := false
+		if node.Status.Conditions != nil {
+			nodeReadyCondition := types.GetCondition(node.Status.Conditions, longhorn.NodeConditionTypeReady)
+			isNodeReady = nodeReadyCondition.Status == longhorn.ConditionStatusTrue
+		}
+		if replica.Spec.FailedAt == "" && replica.Spec.HealthyAt != "" && isNodeReady {
+			healthyCount++
+		}
+	}
+	return healthyCount >= numberOfReplicas, nil
+}
+
+func (vbc *VolumeRebuildingController) isVolumeReplicasRebuilding(vol *longhorn.Volume) (bool, error) {
+	engines, err := vbc.ds.ListVolumeEngines(vol.Name)
+	if err != nil {
+		return false, err
+	}
+	engine, err := vbc.ds.PickVolumeCurrentEngine(vol, engines)
+	if err != nil {
+		return false, err
+	}
+	if engine == nil || engine.Status.ReplicaModeMap == nil {
+		return true, nil
+	}
+
+	healthyCount := 0
+	for _, mode := range engine.Status.ReplicaModeMap {
+		if mode == longhorn.ReplicaModeRW {
+			healthyCount++
+		}
+	}
+
+	if healthyCount < vol.Spec.NumberOfReplicas {
+		return true, nil
+	}
+
+	return false, nil
+}
```

Comment:
If a replica is being rebuilt, the ReplicaMode is WO. Could we check by this mode?

Reply:
Yes, modified to check if a replica is in `WO` mode, we can return `true` directly. Thanks.